libkpod/container_server.go

```diff
@@ -728,7 +728,15 @@ func (c *ContainerServer) HasSandbox(id string) bool {
 func (c *ContainerServer) RemoveSandbox(id string) {
 	c.stateLock.Lock()
 	defer c.stateLock.Unlock()
+	processLabel := c.state.sandboxes[id].ProcessLabel()
 	delete(c.state.sandboxes, id)
+	for _, sb := range c.state.sandboxes {
+		// If process label still in use don't release it
+		if processLabel == sb.ProcessLabel() {
+			return
+		}
+	}
+	label.ReleaseLabel(processLabel)
```

Comment:
The code looks fine to me, I'm just worried about potential performance hits.  Do we know how many sandboxes there'd be in a typical setup?  If it's under a dozen, I'm not too worried, if it's thousands I'm a bit more concerned given the lock.  In that case perhaps we could make a map with the process label as the key and sandbox id as the value?

Reply:
Right, let's have a labes map instead

Reply:
Is the sandbox code in memory?  The call to ReleaseLabel, is just releasing an MCS Label from a map.  The searching through the list would be expensive only if the call to sb.ProcessLabel() causes a read from disk.

Reply:
@runcom @TomSweeneyRedHat Looking at the code, believe this will all happen in memory and never hit the disk.  So even if you had thousands of sandboxes this should not take any time.

Reply:
Dan i was actually talking about the for cycle above, that's gonna be O(N) with N number of sandboxes. If we have a map for labels we can definitely access it with O(1)

Reply:
The question would be whether we can handle the case where multiple sandboxes use the same label.  I guess if we had a map of lists of sandboxes using the process label it would work.  Theoretically we should not be checking the actual label, but the MCS part of the label.

Reply:
We should actually be able to do that

Reply:
Actually, we just need to have a reference counter on labels, map[string]int where we increment it and decrement it should be fine, I have a patch for it if we want to merge this meanwhile

Reply:
SGTM.  Lets get this one in, and then we can look at your improvement.

Reply:
If you still want to apply the patch cause this is not yet merged, here it is:

https://paste.fedoraproject.org/paste/a9RH8vXpmuF-FBuGhkAzJw

Reply:
```diff
diff --git a/libkpod/container_server.go b/libkpod/container_server.go
index 1bc94871..7f710848 100644
--- a/libkpod/container_server.go
+++ b/libkpod/container_server.go
@@ -168,6 +168,7 @@ func New(config *Config) (*ContainerServer, error) {
                        containers:      oci.NewMemoryStore(),
                        infraContainers: oci.NewMemoryStore(),
                        sandboxes:       make(map[string]*sandbox.Sandbox),
+                       processLabels:   make(map[string]int),
                },
                config: config,
        }, nil
@@ -609,6 +610,8 @@ type containerServerState struct {
        containers      oci.ContainerStorer
        infraContainers oci.ContainerStorer
        sandboxes       map[string]*sandbox.Sandbox
+       // process labels reference counter to release them when not used anymore
+       processLabels map[string]int
 }

 // AddContainer adds a container to the container state store
@@ -696,6 +699,7 @@ func (c *ContainerServer) AddSandbox(sb *sandbox.Sandbox) {
        c.stateLock.Lock()
        defer c.stateLock.Unlock()
        c.state.sandboxes[sb.ID()] = sb
+       c.state.processLabels[sb.ProcessLabel()]++
 }

 // GetSandbox returns a sandbox by its ID
@@ -729,6 +733,13 @@ func (c *ContainerServer) RemoveSandbox(id string) {
        c.stateLock.Lock()
        defer c.stateLock.Unlock()
        delete(c.state.sandboxes, id)
+       processLabel := c.state.sandboxes[id].ProcessLabel()
+       c.state.processLabels[processLabel]--
+       labelCounter := c.state.processLabels[processLabel]
+       if labelCounter == 0 {
+               label.ReleaseLabel(processLabel)
+               delete(c.state.processLabels, processLabel)
+       }
 }

 // ListSandboxes lists all sandboxes in the state store
```

Reply:
This looks good, but I think we need to take it one step futher, to make it truly correct.  We should not be concerned about the entire label, but the last component.

If a user created a label
system_u:system_r:container_lockdown_t:s0:c1,c2
and another container was running as 

system_u:system_r:container_t:s0:c1,c2

We still would not want to release the label.

Reply:
@rhatdan, reworked:
```diff
diff --git a/libkpod/container_server.go b/libkpod/container_server.go
index 1bc94871..2009d912 100644
--- a/libkpod/container_server.go
+++ b/libkpod/container_server.go
@@ -5,6 +5,7 @@ import (
 	"fmt"
 	"os"
 	"path/filepath"
+	"strings"
 	"sync"
 	"time"
 
@@ -168,6 +169,7 @@ func New(config *Config) (*ContainerServer, error) {
 			containers:      oci.NewMemoryStore(),
 			infraContainers: oci.NewMemoryStore(),
 			sandboxes:       make(map[string]*sandbox.Sandbox),
+			processLabels:   make(map[string]int),
 		},
 		config: config,
 	}, nil
@@ -609,6 +611,8 @@ type containerServerState struct {
 	containers      oci.ContainerStorer
 	infraContainers oci.ContainerStorer
 	sandboxes       map[string]*sandbox.Sandbox
+	// process labels reference counter to release them when not used anymore
+	processLabels map[string]int
 }
 
 // AddContainer adds a container to the container state store
@@ -691,11 +695,21 @@ func (c *ContainerServer) ListContainers(filters ...func(*oci.Container) bool) (
 	return filteredContainers, nil
 }
 
+// TODO: move this to opencontainers/selinux
+func getSELinuxLevel(label string) string {
+	if len(label) != 0 {
+		con := strings.SplitN(label, ":", 4)
+		return con[3]
+	}
+	return ""
+}
+
 // AddSandbox adds a sandbox to the sandbox state store
 func (c *ContainerServer) AddSandbox(sb *sandbox.Sandbox) {
 	c.stateLock.Lock()
 	defer c.stateLock.Unlock()
 	c.state.sandboxes[sb.ID()] = sb
+	c.state.processLabels[getSELinuxLevel(sb.ProcessLabel())]++
 }
 
 // GetSandbox returns a sandbox by its ID
@@ -729,6 +743,14 @@ func (c *ContainerServer) RemoveSandbox(id string) {
 	c.stateLock.Lock()
 	defer c.stateLock.Unlock()
 	delete(c.state.sandboxes, id)
+	processLabel := c.state.sandboxes[id].ProcessLabel()
+	level := getSELinuxLevel(processLabel)
+	c.state.processLabels[level]--
+	labelCounter := c.state.processLabels[level]
+	if labelCounter == 0 {
+		label.ReleaseLabel(processLabel)
+		delete(c.state.processLabels, processLabel)
+	}
 }
 
 // ListSandboxes lists all sandboxes in the state store
```

Reply:
Looks great,  you will be an SELinux programmer Before you know it.  I will open a PR with go-selinux to add this.

Reply:
context := selinux.NewContext(processLabel)
level := context["level"]

Should work now.